src/DataStructures/VectorImpl.hpp

```diff
@@ -0,0 +1,392 @@
+#else
+// IWYU pragma: no_forward_declare blaze::MapTrait
+#endif  // ((BLAZE_MAJOR_VERSION == 3) && (BLAZE_MINOR_VERSION <= 3))
+// IWYU pragma: no_forward_declare blaze::IsVector
+// IWYU pragma: no_forward_declare blaze::TransposeFlag
+
+/*!
+ * \ingroup DataStructuresGroup
+ * \brief Base class template for various DataVector and related types
+ *
+ * \details The `VectorImpl` class is the generic parent class for vectors
+ * representing collections of related function values, such as `DataVector`s
+ * for contiguous data over a computational domain.
+ *
+ * The `VectorImpl` does not itself define any particular mathematical
+ * operations on the contained values. The `VectorImpl` template class and the
+ * macros defined in `VectorImpl.hpp` assist in the construction of various
+ * derived classes supporting a chosen set of mathematical operations.
+ *
+ * In addition, the equivalence operator `==` is inherited from the underlying
+ * `PointerVector` type, and returns true if and only if the size and contents
+ * of the two compared vectors are equivalent.
+ *
+ * Template parameters:
+ * - `T` is the underlying stored type, e.g. `double`, `std::complex<double>`,
+ *   `float`, etc.
+ * - `VectorType` is the type that should be associated with the VectorImpl
+ *    during mathematical computations. In most cases, inherited types should
+ *    have themselves as the second template argument, e.g.
+ *  ```
+ *  class DataVector : VectorImpl<double, DataVector> {
+ *  ```
+ *  The second template parameter communicates arithmetic type restrictions to
+ *  the underlying Blaze framework. For example, if `VectorType` is
+ *  `DataVector`, then the underlying architecture will prevent addition with a
+ *  vector type whose `ResultType` (which is aliased to its `VectorType`) is
+ *  `ModalVector`.  Since `DataVector`s and `ModalVector`s represent data in
+ *  different spaces, we wish to forbid several operations between them. This
+ *  vector-type-tracking through an expression prevents accidental mixing of
+ *  vector types in math expressions.
+ *
+ * \note
+ * - If created with size 0, then  `data()` will return `nullptr`
+ * - If either `SPECTRE_DEBUG` or `SPECTRE_NAN_INIT` are defined, then the
+ *   `VectorImpl` is default initialized to `signaling_NaN()`. Otherwise, the
+ *   vector is filled with uninitialized memory for performance.
+ */
+template <typename T, typename VectorType>
+class VectorImpl
+    : public PointerVector<T, blaze::unaligned, blaze::unpadded,
+                           blaze::defaultTransposeFlag, VectorType> {
+  /// \cond HIDDEN_SYMBOLS
+  // the `static_assert` requires a member function around it, as it then
+  // gains access to the fully defined type it is a member of, and therefore
+  // can test the resulting class.
+  static constexpr void private_asserts() noexcept {
+    static_assert(
+        std::is_nothrow_move_constructible<VectorImpl<T, VectorType>>::value,
+        "Missing move semantics");
```

Comment:
This doesn't work.  I changed `is_nothrow_move_constructible` to `is_fundamental` and it still compiled.

Reply:
hmm.. I suppose that makes sense, because it's a template now. I can move it to the derived classes, although now that I'm looking at it more closely, it's not clear to me that this is better suited as an assert in each of the classes, or as an assert in the `DataVector` etc. tests.

Reply:
I think the only reason it exists is because we found some bug in GCC5 or older. I'm fine with just removing it.